File: Week04/LeetCode-0349-2.py
class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        O(m+n) algorithm with O(1) extra space using a LUT
        """
        maxNum = 1000
        lut1 = [0]*(maxNum+1)
        for num in nums1: lut1[num] = 1

        result = []
        for num in nums2:
            if lut1[num] == 1:
                result.append(num)
                lut1[num] = 0  # To avoid duplicates in the result
        return result        
        

File: Week04/LeetCode-0349-2.cpp
class Solution {
public:
    ///-----------------------------------------------------------------
    /// Here is an O(n+m) solution that uses O(n+m) extra space
    ///
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        /// O(m+n) algorithm with O(1) extra space using a LUT
        int maxNum = 1000;
        vector<int> lut1(maxNum+1, 0);
        
        for (auto num: nums1) lut1[num] = 1;

        vector<int> result;
        for (auto num: nums2){
            if (lut1[num] == 1){
                result.push_back(num);
                lut1[num] = 0;  // To avoid duplicates in the result
            } // end-if
        } //end-for
        
        return result;                
    } //end-intersection
};